src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -322,6 +380,14 @@ def main():
         conf.pools.close()
         conf.pools.join()
 
+        # create a softlink from the relevant directory under collectionsdata
+        if collection_source_dir:
+            for upa in upas:
+                upa_dir = os.path.join(output_dir, upa)
+                csd_upa_dir = os.path.join(collection_source_dir, upa)
+                if not os.path.islink(csd_upa_dir):
```

Comment:
My current logic is still not correct.. Need to put into a function that return a bool

Reply:
Updated!

Reply:
Hmm, the current logic of deleting anything in the `collectionssource` directory if it's not the correct link makes me a bit nervous. This is not something that should normally happen since only this script should be making changes to the directory hierarchy, so I'm thinking if that there's something blocking the link operation the script should just throw an error and let the user resolve it.

Since the script now has "pick up where you left off" mechanics and, generally speaking, this shouldn't happen often if ever, that seems safer

Reply:
Also you might was well move the actual symlink operation into _create_softlink

Reply:
Updated. I think I am making it more complicated that it should be. May just below logic should be enough?

```
try:
    os.symlink(upa_dir, csd_upa_dir, target_is_directory=True)
except error as e:
    raise ValueError("two directories are not successfully linked") from e
```

Reply:
the current implementation still misses the case where csd_upa_dir is a file. I'd do something like this (psuedocode):

```
if src.exists():
    if src.is_link() and src.link() == target:
        return
    raise ValueError(f"{src} already exists and does not link to {target} as expected")
target.link(src)  # this is a race condition, but if src is created after the exists() call it'll just throw an error
```

Reply:
:+1: